### ISSUE TITLE
Virtual ref not populated correctly

### ISSUE THREAD

**Issue by robert52** [Automattic/mongoose#4618]
Hi,

I'm trying to use virtual ref in a One-One way. Example:

```
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const ObjectId = Schema.ObjectId;

const ASchema = new Schema({
  name: { type: String }
});

const BSchema = new Schema({
  name: { type: String }
  a_id: { type: ObjectId }
}, {
  toObject: { virtuals: true },
  toJSON:   { virtuals: true }
});

BSchema.virtual('a', {
  ref: 'A',
  localField: 'a_id',
  foreignField: '_id'
});

// ... compile models

B.findById('<some_id>').populate('a').exec((err, b) => {
   if (err) throw(err);

   console.log(b);
});
```

Output:

```
{
  _id: 57fb99b22780e94832ef3f42,
  name: 'foo',
  a_id: 57fb99b22780e94832ef4a31,
  a: undefined,
  id: '57fb99b22780e94832ef3f42'
}
```

In this case, I think `a` should be a single document.

Running a populate on the doc:

```
B.findById('<some_id>', (err, b) => {
   if (err) throw(err);

   b.populate('a', () => {
     console.log(b);
   });
});
```

Output:

```
{
  _id: 57fb99b22780e94832ef3f42,
  name: 'foo',
  a_id: 57fb99b22780e94832ef4a31,
  a: [{
    _id: 57fb99b22780e94832ef4a31,
    name: 'bar'
  }],
  id: '57fb99b22780e94832ef3f42'
}
```

Returning an `Array` from the populate seems a little bit strange to me. There should be an option for the `virtual ref`, if it should return a single document or multiple documents.

Using mongoose version: `4.6.3`

Thanks,
Robert.

**Comment by TrejGun** [comment]
try
```
justOne: true
```
option

**Comment by SergioDonati** [comment]
I have the same problem, add justOne options not resolve.

Using 

```javascript
B.findById('<some_id>').populate('a').exec((err, b) => {
   // b.a is undefined here
});
```

instead when

```javascript
B.findById('<some_id>', (err, b) => {
    b.populate('a', () => {
        // b.a is defined with A model object
    });
});
```

Using mongoose version: `4.6.4`
mongoDB version: `3.2.9`
node version: `6.8.1`

Thanks

**Comment by SergioDonati** [comment]
I make an unit test with mocha

[on gist](https://gist.github.com/SergioDonati/f47f1e68f80f9d46fb0030846fcbc1d2)

**Comment by robert52** [comment]
@SergioDonati thanks for the gist

@TrejGun it seems something else is the problem, I did some debugging on the source code.
The `populate()` function from `model.js` is working fine and data for `refs` are retrieved correctly. Somehow the data is lost, I've hit a dead end at the moment, because I don't know where exactly this happens, as it seems the data is assigned correctly when setting virtuals and other paths. The final model instance returned in the callback is different and has the `virtual ref` as `undefined`

**Comment by vkarpov15** [comment]
Should be fixed in 4.6.6, weird issue with document init swallowing the virtual. Make sure you also use the `justOne` option if you want a single doc :+1:

**Comment by TrejGun** [comment]
fixed, thanks
